Add turing_req_id in requests made to Turing Router components

deadlycoconuts opened the pull request:
## Context
The `turing_req_id` is a unique identifier attached to every response header for each request sent to a Turing Router. However this header is only added just before the response is returned to the user and could instead be sent together as part of requests sent to the other Turing Router components (e.g. enricher, experiments engine, routes, ensembler), allowing users to capitalise on this identifier for other uses (e.g. using the id as a randomisation key).

This PR thus introduces a slight modification to the Turing Router, by adding the `turing_req_id` generated whenever each request is received to all the subsequent request headers that get sent to the other Turing Router components. 

Additionally, this PR replaces the deprecated `ioutil.ReadAll` function with the new `io.ReadAll` function that gets used in various parts of the Turing Router.

## Main Modifications 
- `engines/router/missionctl/handlers/http_handler.go` - Add `turing_req_id` to requests sent to Turing Router components
- `engines/router/missionctl/handlers/http_handler_test.go` - Update unit tests affected by the aforementioned change
- `docs/*` - Add documentation to Turing Router components affected by the aforementioned change 
- `sdk/samples/router/create_router_with_pyfunc_ensembler.py` - Add sample to demonstrate how the `turing_req_id` header can be accessed in a Pyfunc ensembler

krithika369 commented on engines/router/missionctl/handlers/http_handler.go:
```diff
@@ -187,7 +191,6 @@ func (h *httpHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
 	for key := range resp.Header() {
 		rw.Header().Set(key, resp.Header().Get(key))
 	}
-	rw.Header().Set(turingReqIDHeaderKey, turingReqID)
```
We should still set the `turingReqIDHeaderKey` in the response header as if not, it will break the current behavior.

deadlycoconuts replied:
Oops good catch; I've added it back again

krithika369 reviewed (commented):
Thanks for the quick fix - it LGTM. I realise you haven't requested review on the PR yet, so likely it's WIP. But one comment - can we test and document (in the user docs / SDK examples) the following:
* How the Turing-Req-Id can be used as a randomization key with the experiment engine (there may be some room to simplify the configuration of this in the router form, but we can defer that for the time being)
* How the Turing-Req-Id may be accessed in the Pyfunc ensembler

deadlycoconuts commented on sdk/samples/router/create_router_with_pyfunc_ensembler.py:
```diff
@@ -55,7 +55,10 @@ def ensemble(
             return predictions[input["version"]]["data"]["value"]
         elif "Version" in kwargs["headers"]:
-            return predictions[kwargs["headers"]["Version"]]["data"]["value"]
+            if hash(kwargs["headers"]["Turing-Req-Id"]) % 2 == 0:  # If the hash of the turing request id is even
+                return predictions[kwargs["headers"]["Version"]]["data"]["value"]
+            else:
+                return 0
```
Added this additional if/else condition to demonstrate how the `Turing-Req-Id` header can be accessed in a Pyfunc ensembler.

deadlycoconuts commented:
Okay, I've tested the presence of the header in calls made the Router routes and added additional documentation for the enricher/experiment engine/routes/ensembler sections to mention the inclusion of the `turing_req_id` in the request headers sent to these components.

I've also modified the sample pyfunc ensembler to demonstrate how the new header can be accessed in the `ensemble` method.

krithika369 commented on docs/how-to/create-a-router/configure-enricher.md:
```diff
@@ -10,7 +10,12 @@ The Turing enricher has the ability to perform arbitrary transformations on the
+Turing will deploy specified Docker image as a pre-processor and will send original request to it for enrichment. 
+
+The request to the enricher is constructed from the request headers to the Router, and an identifier `turing_req_id` 
+that is uniquely assigned to each request received by the Router.
```
`that is uniquely assigned to each request received by the Router` -> `that is uniquely assigned to each request "is" received by the "Enricher"`  ?

Also, isn't it `Turing-Req-ID` instead of `turing_req_id` ?

deadlycoconuts replied:
Oh yeah, that's right; it should be `Turingn-Req-Id`

krithika369 commented on docs/how-to/create-a-router/configure-ensembler.md:
```diff
@@ -26,7 +26,14 @@ It is not possible to select as the fallback response a route that has traffic r
+ensembling - the original request, responses from all routes, and the treatment configuration (if an Experiment 
+Engine is selected, in the Configure Experiment Engine step). The ensembler's request headers will contain the original 
+request headers sent to Turing, merged with the enricher's response headers (if there are duplicates, the value in 
+the enricher's response headers will take precedence), and an identifier `turing_req_id` that is uniquely assigned to each request received by the Router. 
```
Same comments as with `docs/how-to/create-a-router/configure-enricher.md` - missing "is" and `Router` -> `Ensembler`; `turing_req_id` -> `Turing-Req-ID`.

krithika369 commented on docs/how-to/create-a-router/configure-ensembler.md:
```diff
@@ -57,20 +64,26 @@ Configure the resources required for the ensembler. There are 3 required inputs,
+original request, responses from all routes, and the treatment configuration (if an Experiment
+Engine is selected, in the Configure Experiment Engine step). The ensembler's request headers will contain the original
+request headers sent to Turing, merged with the enricher's response headers (if there are duplicates, the value in
+the enricher's response headers will take precedence), and an identifier `turing_req_id` that is uniquely assigned to each request received by the Router. 
```
Same comments as above.

krithika369 commented on docs/how-to/create-a-router/configure-experiment-engine.md:
```diff
@@ -4,6 +4,10 @@
 {% endhint %}
 
+The request sent to the experiment engine is constructed from the request headers to the Router (including an
+identifier `turing_req_id` that is uniquely assigned to each request received by the Router), and either the request 
```
`Turing-Req-ID` here too ?

krithika369 commented on docs/how-to/create-a-router/configure-routes.md:
```diff
@@ -2,6 +2,10 @@
 Routes are an essential part of your Turing Router setup. Each route is defined by its ID and the endpoint. A route can be your deployed ML model, exposed via HTTP interface, or an arbitrary non-ML web-service. Each router should be configured with at least one route and each route's ID should be unique among other routes of this router.
 
+The request sent to the each route is constructed from the request headers to the Router (including an
+identifier `turing_req_id` that is uniquely assigned to each request received by the Router), and either the request
```
Same comment as above.

krithika369 reviewed (approved):
Left some comments on the docs - the rest LGTM. Thanks ! 🚀

deadlycoconuts commented:
Thanks for the comments! I'll be merging this now!
